pkg/e2/v1beta1/node.go

```diff
@@ -0,0 +1,199 @@
+// SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetworking.org>
+//
+// SPDX-License-Identifier: LicenseRef-ONF-Member-1.0
+
+package e2
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/uuid"
+	e2api "github.com/onosproject/onos-api/go/onos/e2t/e2/v1beta1"
+	"github.com/onosproject/onos-lib-go/pkg/env"
+	"github.com/onosproject/onos-lib-go/pkg/errors"
+	"google.golang.org/grpc"
+	"io"
+	"sync"
+)
+
+// NodeID is an E2 node identifier
+type NodeID string
+
+// Node is an E2 node
+type Node interface {
+	// ID is the node identifier
+	ID() NodeID
+
+	// Subscribe creates a subscription from the given SubscriptionDetails
+	// The Subscribe method will block until the subscription is successfully registered.
+	// The context.Context represents the lifecycle of this initial subscription process.
+	// Once the subscription has been created and the method returns, indications will be written
+	// to the given channel.
+	// If the subscription is successful, a subscription.Context will be returned. The subscription
+	// context can be used to cancel the subscription by calling Close() on the subscription.Context.
+	Subscribe(ctx context.Context, sub *e2api.Subscription, indCh chan<- e2api.Indication, errCh chan<- error) error
+
+	// Control creates and sends a E2 control message and awaits the outcome
+	Control(ctx context.Context, message *e2api.ControlMessage) (*e2api.ControlOutcome, error)
```

Comment:
The only problem that I see with this interface is returning just control outcome. We don't know yet what will be the future of Control outcome based on the current discussions going on in O-RAN but it is likely to be removed and the E2 node returns just a Control Ack. It would be better to keep the pattern to be Control Request and Control Response (or Control Ack) and outcome will be part of it which allows us to remove it if we want to.

Reply:
@adibrastegarnia the reason the API does not accept the top-level request messages or return the top level response messages is because it hides the headers from user code. This allows the code populating headers to be handled internally. 

I don’t think we can worry about O-RAN changing the API too much. If changes are made to the control outcome they’re going to break the API either way because user code relying those fields will have to change. At least introducing versioning into our own APIs will allow us to handle those breaking changes gracefully. If they remove the control outcome, we’ll just have to bump the ONOS API to a new version to handle the breaking changes in E2AP.

But I don’t think we ever really intended to have onos-api messages as part of the public SDK API. I just used the onos-api messages in the SDK interfaces because I knew they’d change so it wasn’t worth investing too much energy creating a clean API. Maybe we do need to address that now, and then the public SDK interfaces may not need to be versioned after all. We’d still need to version the internal clients used to talk to different versions of onos-api services, but we could probably negotiate the API version for each service behind a single SDK interface. So, the SDK would define its own structs and interfaces for defining control requests and subscriptions, and those would be compiled to a specific API version based on the version available within the cluster.

Reply:
I agree with you overall with defining structs that we can add whatever fields we need to those structs but the problem is using just api.Control outcome doesn't help us to see if the request is failed or not. This is important even with current version of E2AP Control API. Control Failure also has a Control outcome at the moment. Control outcome is the service model specefic part of a Control Ack or Control failure message. If we can define new messages (whatever we would like to call it in SDK) and wrap control outcome and its status (failed or success) then would be great and I think it is essential for upcoming DT trial.

Reply:
Unless you want to use error to distinguish them from each other.

Reply:
Ahh... yes. I was aware of that. One thing I forgot to mention in the original description was that I intended to start adding more rich errors to the API/SDK. gRPC supports an [extended error model](https://grpc.io/docs/guides/error/) that allows arbitrary proto messages to be returned as part of the error. The [Go client supports it](https://jbrandhorst.com/post/grpc-errors/), as do most of the clients in other popular languages. So the idea here is to return richer details as part of the gRPC error response. This is the purpose of the `Error` message added in this PR, which includes all the E2AP error causes. I haven’t handled it yes in this PR, but the failure outcome should become part of the error details and should be retrievable within the SDK using the error API.